src/org/rust/lang/core/psi/util/RustModExtensions.kt

```diff
@@ -0,0 +1,61 @@
+package org.rust.lang.core.psi.util
+
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RustCompositeElement
+import org.rust.lang.core.psi.RustCrateItem
+import org.rust.lang.core.psi.RustModItem
+
+
+private val RustModItem.modDir: PsiDirectory?
+    get() {
+        val parent = parentModInFile ?: return containingFile.parent
+
+        // FIXME(matklad): mod name should be not null, should we rearange our rules for file/mod?
+        val name = this.name ?: return null
+
+        return parent.modDir?.findSubdirectory(name)
+    }
+
+
+val RustCompositeElement.parentModInFile: RustModItem?
+    get() = parentOfType<RustModItem>()
+
+private val RustModItem.isCrateRoot: Boolean
+    get() = parentModInFile == null &&
+        (containingFile.name == "main.rs" || containingFile.name == "lib.rs")
+
+private val RustModItem.ownsDirectory: Boolean
+    get() = parentModInFile != null || // any inline nested module owns a directory
+        containingFile.name == "mod.rs" ||
+        isCrateRoot
+
+
+sealed class ChildModFile {
+    class NotFound : ChildModFile()
+    class Found(val mod: RustCrateItem) : ChildModFile()
+    class Ambigious(val paths: Collection<PsiFile>) : ChildModFile()
+}
+
+fun RustModItem.submodulePath(modName: String): ChildModFile {
+    if (!ownsDirectory) {
+        return ChildModFile.NotFound()
+    }
+
+    val dir = modDir
+    val dirMod = dir?.findSubdirectory(modName)?.findFile("mod.rs")
+    val fileMod = dir?.findFile("$modName.rs")
+
+    val file = when {
+        dirMod != null && fileMod != null -> return ChildModFile.Ambigious(listOf(dirMod, fileMod))
+        dirMod != null                    -> dirMod
+        fileMod != null                   -> fileMod
+        else                              -> return ChildModFile.NotFound()
```

Comment:
I think better add     
```Kotlin
val variants = listOf(dirMod, fileMod).filterNotNull()
```
and work with it. This conditions a little bit bulky.